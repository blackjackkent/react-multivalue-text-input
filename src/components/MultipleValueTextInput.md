### Basic example:

```js
<MultipleValueTextInput
	onItemAdded={(item, allItems) => console.log(`Item added: ${item}`)}
	onItemDeleted={(item, allItems) => console.log(`Item removed: ${item}`)}
	label="Items"
	name="item-input"
	placeholder="Enter whatever items you want; separate them with COMMA or ENTER."
/>
```

### Custom delete button:

```js
<MultipleValueTextInput
	onItemAdded={(item, allItems) => console.log(`Item added: ${item}`)}
	onItemDeleted={(item, allItems) => console.log(`Item removed: ${item}`)}
	label="Items"
	name="item-input"
	placeholder="Enter whatever items you want; separate them with COMMA or ENTER."
	deleteButton={<span>(delete)</span>}
/>
```

### Populate initial values

```js
<MultipleValueTextInput
	onItemAdded={(item, allItems) => console.log(`Item added: ${item}`)}
	onItemDeleted={(item, allItems) => console.log(`Item removed: ${item}`)}
	label="Items"
	name="item-input"
	placeholder="Enter whatever items you want; separate them with COMMA or ENTER."
	values={['default value', 'another default value']}
/>
```

### Custom delimiter

```js
<MultipleValueTextInput
	onItemAdded={(item, allItems) => console.log(`Item added: ${item}`)}
	onItemDeleted={(item, allItems) => console.log(`Item removed: ${item}`)}
	label="Items"
	name="item-input"
	submitKeys={[' ']}
	placeholder="Enter whatever items you want; separate them with SPACE."
/>
```

### Add field content on blur

```js
<MultipleValueTextInput
	onItemAdded={(item, allItems) => console.log(`Item added: ${item}`)}
	onItemDeleted={(item, allItems) => console.log(`Item removed: ${item}`)}
	label="Items"
	name="item-input"
	shouldAddOnBlur={true}
	placeholder="Enter whatever items you want; deselect/blur the input to add item and start a new one."
/>
```

### Populate values dynamically

```js
const [values, setValues] = React.useState([]);
const populateValuesDynamically = () => {
	setValues(['test1', 'test2', 'test3']);
};

<div>
	<button type="button" onClick={populateValuesDynamically}>
		Populate Some Data
	</button>
	<MultipleValueTextInput
		label="Items"
		name="item-input"
		placeholder="Enter whatever items you want; separate them with COMMA or ENTER. Or click the button to populate dynamically."
		onItemAdded={(_, resultItems) => setValues(resultItems)}
		onItemDeleted={(_, resultItems) => setValues(resultItems)}
		values={values}
	/>
</div>;
```
