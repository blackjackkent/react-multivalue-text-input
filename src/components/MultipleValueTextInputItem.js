import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
	value: PropTypes.string.isRequired,
	handleItemRemove: PropTypes.func.isRequired,
	deleteButton: PropTypes.node.isRequired
};

const MultipleValueTextInputItem = (props) => {
	const { value, handleItemRemove, deleteButton } = props;
	return (
		<span className="multiple-value-text-input-item" role="listitem">
			<span className="multiple-value-text-input-item-value" data-testid="value">
				{value}
			</span>{' '}
			<span
				className="multiple-value-text-input-item-delete-button"
				data-value={value}
				tabIndex="-1"
				role="button"
				onKeyPress={() => handleItemRemove(value)}
				onClick={() => handleItemRemove(value)}
			>
				{deleteButton}
			</span>
		</span>
	);
};

MultipleValueTextInputItem.propTypes = propTypes;
export default MultipleValueTextInputItem;
