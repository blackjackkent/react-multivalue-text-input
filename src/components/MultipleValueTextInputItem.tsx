import React from 'react';
import PropTypes from 'prop-types';
import { MultipleValueTextInputItemProps } from '../types';
import styles from './MultipleValueTextInputItem.module.css';

const propTypes = {
	value: PropTypes.string.isRequired,
	handleItemRemove: PropTypes.func.isRequired,
	deleteButton: PropTypes.node.isRequired
};

const MultipleValueTextInputItem = (props: MultipleValueTextInputItemProps) => {
	const { value, handleItemRemove, deleteButton } = props;
	return (
		<span className={`multiple-value-text-input-item ${styles.inputItem}`} role="listitem">
			<span data-testid="value">{value}</span>{' '}
			<span
				className={`multiple-value-text-input-item-button ${styles.deleteButton}`}
				data-value={value}
				tabIndex={-1}
				role="button"
				onKeyUp={() => handleItemRemove(value)}
				onClick={() => handleItemRemove(value)}
			>
				{deleteButton}
			</span>
		</span>
	);
};

MultipleValueTextInputItem.propTypes = propTypes;
export default MultipleValueTextInputItem;
