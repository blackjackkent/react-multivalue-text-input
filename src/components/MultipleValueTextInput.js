import React, { Component } from 'react';
import PropTypes from 'prop-types';
import '../styles/styles.scss';
import MultipleValueTextInputItem from './MultipleValueTextInputItem';

const propTypes = {
	/** Any values the input's collection should be prepopulated with. */
	values: PropTypes.arrayOf(PropTypes.string),
	/** Method which should be called when an item is added to the collection */
	onItemAdded: PropTypes.func.isRequired,
	/** Method which should be called when an item is removed from the collection */
	onItemDeleted: PropTypes.func.isRequired,
	/** Label to be attached to the input, if desired */
	label: PropTypes.string,
	/** Name attribute for the input */
	name: PropTypes.string.isRequired,
	/** Placeholder attribute for the input, if desired */
	placeholder: PropTypes.string,
	/** ASCII charcode for the keys which should
	 * trigger an item to be added to the collection (defaults to comma (44) and Enter (13))
	 */
	charCodes: PropTypes.arrayOf(PropTypes.number),
	/** JSX or string which will be used as the control to delete an item from the collection */
	deleteButton: PropTypes.node,
	/** Whether or not the blur event should trigger the added-item handler */
	shouldAddOnBlur: PropTypes.bool,
	/** Custom class name for the input element */
	className: PropTypes.string,
	/** Custom class name for the input label element */
	labelClassName: PropTypes.string
};

const defaultProps = {
	placeholder: '',
	charCodes: [13, 44],
	deleteButton: (<span>&times;</span>),
	values: [],
	label: '',
	shouldAddOnBlur: false,
	className: '',
	labelClassName: ''
};
/**
 * A text input component for React which maintains and displays a collection
 * of entered values as an array of strings.
 */
class MultipleValueTextInput extends Component {
	constructor(props) {
		super(props);
		this.state = {
			values: props.values,
			value: ''
		};
		this.handleKeypress = this.handleKeypress.bind(this);
		this.handleValueChange = this.handleValueChange.bind(this);
		this.handleItemAdd = this.handleItemAdd.bind(this);
		this.handleItemRemove = this.handleItemRemove.bind(this);
		this.handleBlur = this.handleBlur.bind(this);
	}
	handleKeypress(e) {
		const { onItemAdded, charCodes } = this.props;
		// 13: Enter, 44: Comma
		if (charCodes.includes(e.charCode)) {
			e.preventDefault();
			this.handleItemAdd(e.target.value, onItemAdded);
		}
	}
	handleValueChange(e) {
		this.setState({ value: e.target.value });
	}
	handleItemAdd(value, onItemAdded) {
		if (this.state.values.includes(value) || !value) {
			this.setState({ value: '' });
			return;
		}
		const newValues = this.state.values.concat(value);
		this.setState({
			values: newValues,
			value: ''
		});
		onItemAdded(value, newValues);
	}
	handleItemRemove(value) {
		const currentValues = this.state.values;
		const newValues = currentValues.filter(v => v !== value);
		this.props.onItemDeleted(value, newValues);
		this.setState({ values: newValues });
	}
	handleBlur(e) {
		const { onItemAdded, shouldAddOnBlur } = this.props;
		// 13: Enter, 44: Comma
		if (shouldAddOnBlur) {
			e.preventDefault();
			this.handleItemAdd(e.target.value, onItemAdded);
		}
	}
	render() {
		const {
			placeholder,
			label,
			name,
			deleteButton,
			onItemAdded,
			onItemDeleted,
			className,
			labelClassName,
			charCodes,
			values: InputValues,
			shouldAddOnBlur,
			...forwardedProps
		} = this.props;
		const values = this.state.values && this.state.values.length
			? this.state.values
			: this.props.values;
		const valueDisplays = values.map(v => (
			<MultipleValueTextInputItem
				value={v}
				key={v}
				deleteButton={deleteButton}
				handleItemRemove={this.handleItemRemove}
			/>));
		return (
			<div className="multiple-value-text-input">
				<label htmlFor={name} className={labelClassName}>{label}
					<div className="multiple-value-text-input-item-container">
						{values.length > 0 && <p>{valueDisplays}</p>}
					</div>
					<input
						name={name}
						placeholder={placeholder}
						value={this.state.value}
						type="text"
						onKeyPress={this.handleKeypress}
						onChange={this.handleValueChange}
						onBlur={this.handleBlur}
						className={className}
						{...forwardedProps}
					/>
				</label>
			</div>
		);
	}
}

MultipleValueTextInput.propTypes = propTypes;
MultipleValueTextInput.defaultProps = defaultProps;
export default MultipleValueTextInput;
