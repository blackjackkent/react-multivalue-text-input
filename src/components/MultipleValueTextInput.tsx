import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styles from './MultipleValueTextInput.module.css';
import { MultipleValueTextInputProps } from '../types';
import MultipleValueTextInputItem from './MultipleValueTextInputItem';

const propTypes = {
	/** Any values the input's collection should be prepopulated with. */
	values: PropTypes.arrayOf(PropTypes.string),
	/** Method which should be called when an item is added to the collection */
	onItemAdded: PropTypes.func.isRequired,
	/** Method which should be called when an item is removed from the collection */
	onItemDeleted: PropTypes.func.isRequired,
	/** Label to be attached to the input, if desired */
	label: PropTypes.string,
	/** Name attribute for the input */
	name: PropTypes.string.isRequired,
	/** Placeholder attribute for the input, if desired */
	placeholder: PropTypes.string,
	/** ASCII charcode for the keys which should
	 * trigger an item to be added to the collection (defaults to comma (44) and Enter (13))
	 */
	submitKeys: PropTypes.arrayOf(PropTypes.string),
	/** JSX or string which will be used as the control to delete an item from the collection */
	deleteButton: PropTypes.node,
	/** Whether or not the blur event should trigger the added-item handler */
	shouldAddOnBlur: PropTypes.bool,
	/** Custom class name for the input element */
	className: PropTypes.string,
	/** Custom class name for the input label element */
	labelClassName: PropTypes.string
};

/**
 * A text input component for React which maintains and displays a collection
 * of entered values as an array of strings.
 */
function MultipleValueTextInput({
	placeholder = '',
	label = '',
	name,
	deleteButton = <span>&times;</span>,
	onItemAdded = () => null,
	onItemDeleted = () => null,
	className = '',
	labelClassName = '',
	submitKeys = ['Enter', ','],
	values: initialValues = [],
	shouldAddOnBlur,
	...forwardedProps
}: MultipleValueTextInputProps) {
	const [values, setValues] = useState(initialValues);
	const [value, setValue] = useState('');
	const nonCharacterKeyLabels: string[] = ['Enter','Tab']
	const delimiters: string[] = submitKeys.filter( ( element ) => !nonCharacterKeyLabels.includes( element ) );
	const handleValueChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		setValue(e.currentTarget.value);
	};
	const handleItemAdd = (addedValue: string) => {
		if (values.includes(addedValue) || !addedValue) {
			setValue('');
			return;
		}
		const newValues = values.concat(addedValue);
		setValues(newValues);
		setValue('');
		onItemAdded(value, newValues);
	};
	const handleItemsAdd = (addedValues: string[]) => {
		let uniqueValues: string[] = []
		addedValues.forEach((addedValue) => {
			if (addedValue && !values.includes(addedValue) && !uniqueValues.includes(addedValue)) {
				uniqueValues = uniqueValues.concat(addedValue)
			}
		})
		if (uniqueValues.length > 0) {
			const newValues = values.concat(uniqueValues)
			setValues(newValues)
			setValue('')
			uniqueValues.forEach((addedValue) => {
				onItemAdded(addedValue, newValues)
			})
		} else {
			setValue('')
		}
	};
	const handleItemRemove = (removedValue: string) => {
		const currentValues = values;
		const newValues = currentValues.filter((v) => v !== removedValue);
		onItemDeleted(removedValue, newValues);
		setValues(newValues);
	};

	const handleKeypress = (e: React.KeyboardEvent<HTMLInputElement>) => {
		// Defaults: Enter, Comma (e.key === 'Enter' or ',')
		if (submitKeys.includes(e.key)) {
			e.preventDefault();
			handleItemAdd(e.currentTarget.value);
		}
	};
	const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
		if (shouldAddOnBlur) {
			e.preventDefault();
			handleItemAdd(e.target.value);
		}
	};

	const splitMulti = (str: string) => {
		const tempChar = delimiters[0] // We can use the first token as a temporary join character
		let result: string = str
		for (let i = 1; i < delimiters.length; i+=1) {
			result = result.split(delimiters[i]).join(tempChar) // Handle scenarios where pasted text has more than one submitKeys in it
		}
		return result.split(tempChar)
	}

	const handlePaste = (e: React.ClipboardEvent<HTMLInputElement>) => {
		const pastedText = e.clipboardData.getData('text/plain')
		let areSubmitKeysPresent = false
		delimiters.forEach((delimiter) => {
			if(pastedText.includes(delimiter)) {
				areSubmitKeysPresent = true
			}
		})
		if (areSubmitKeysPresent) {
			const splitTerms = splitMulti(pastedText)
			if (splitTerms.length > 0) {
				e.preventDefault()
				handleItemsAdd(splitTerms)
			}
		}
	}

	const valueDisplays = values.map((v) => (
		<MultipleValueTextInputItem
			value={v}
			key={v}
			deleteButton={deleteButton}
			handleItemRemove={handleItemRemove}
		/>
	));
	return (
		<div className="multiple-value-text-input" role="form">
			<label
				htmlFor={name}
				className={`${labelClassName} ${styles.inputLabel}`}
				data-testid="label"
			>
				{label}
				<div className="multiple-value-text-input-item-container">
					{values.length > 0 && <p role="list">{valueDisplays}</p>}
				</div>
				<input
					aria-label={label}
					name={name}
					placeholder={placeholder}
					value={value}
					type="text"
					onKeyPress={handleKeypress}
					onChange={handleValueChange}
					onPaste={handlePaste}
					onBlur={handleBlur}
					className={`${className} ${styles.inputElement}`}
					{...forwardedProps}
				/>
			</label>
		</div>
	);
}

MultipleValueTextInput.propTypes = propTypes;
export default MultipleValueTextInput;
