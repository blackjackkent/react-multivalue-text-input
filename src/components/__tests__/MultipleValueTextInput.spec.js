import React from 'react';
import { render, screen } from '@testing-library/react';
import MultipleValueTextInput from '../MultipleValueTextInput';
import MultipleValueTextInputItem from '../MultipleValueTextInputItem';

const createTestProps = (propOverrides) => ({
	// common props
	onItemAdded: jest.fn(),
	onItemDeleted: jest.fn(),
	name: 'test-input',
	...propOverrides
});

describe('rendering', () => {
	it('should render valid snapshot with default props', () => {
		const props = createTestProps();
		const input = shallow(<MultipleValueTextInput {...props} />);
		expect(input).toMatchSnapshot();
	});
	it('should render valid snapshot with passed props', () => {
		const props = createTestProps({ label: 'Test Label', placeholder: 'Test Placeholder' });
		const input = shallow(<MultipleValueTextInput {...props} />);
		expect(input).toMatchSnapshot();
	});
	it('should render with default props', () => {
		const props = createTestProps();
		const input = shallow(<MultipleValueTextInput {...props} />);
		expect(input.find(MultipleValueTextInputItem)).toHaveLength(0);
		expect(input.find('label')).toHaveText('');
		expect(input.find('input')).toHaveProp('name', 'test-input');
		expect(input.find('input')).toHaveProp('placeholder', '');
		expect(input.find('input')).toHaveValue('');
	});
	it('should render with passed props', () => {
		const props = createTestProps({ label: 'Test Label', placeholder: 'Test Placeholder' });
		const input = shallow(<MultipleValueTextInput {...props} />);
		expect(input.find(MultipleValueTextInputItem)).toHaveLength(0);
		expect(input.find('label')).toHaveText('Test Label');
		expect(input.find('input')).toHaveProp('name', 'test-input');
		expect(input.find('input')).toHaveProp('placeholder', 'Test Placeholder');
		expect(input.find('input')).toHaveValue('');
	});
	it('should render items', () => {
		const props = createTestProps({ values: ['1', '2', '3'] });
		const input = shallow(<MultipleValueTextInput {...props} />);
		expect(input.find(MultipleValueTextInputItem)).toHaveLength(3);
	});
	it('should forward props to input', () => {
		const props = createTestProps({ 'data-test': 'test' });
		const input = shallow(<MultipleValueTextInput {...props} />);
		expect(input.find('input')).toHaveProp('data-test', 'test');
	});
	it('should forward props to item', () => {
		const props = createTestProps({ values: ['1'], deleteButton: 'blah' });
		const input = shallow(<MultipleValueTextInput {...props} />);
		const item = input.find(MultipleValueTextInputItem).first();
		expect(item).toHaveValue('1');
		expect(item).toHaveProp('deleteButton', 'blah');
	});
});

describe('behavior', () => {
	it('should set value on change', () => {
		const props = createTestProps();
		const input = shallow(<MultipleValueTextInput {...props} />);
		input.find('input').simulate('change', { target: { value: 'abc' } });
		expect(input.find('input')).toHaveProp('value', 'abc');
	});
	it('should handle default char codes to add values', () => {
		const onItemAdd = jest.fn();
		const enterEvent = { preventDefault: () => {}, charCode: 13, target: { value: '1' } };
		const commaEvent = { preventDefault: () => {}, charCode: 44, target: { value: '2' } };
		const otherEvent = { preventDefault: () => {}, charCode: 10, target: { value: '3' } };
		const props = createTestProps({ onItemAdded: onItemAdd });
		const input = shallow(<MultipleValueTextInput {...props} />);
		input.find('input').simulate('keyPress', enterEvent);
		input.find('input').simulate('keyPress', commaEvent);
		input.find('input').simulate('keyPress', otherEvent);
		expect(onItemAdd).toHaveBeenCalledTimes(2);
	});
	it('should handle custom char codes to add values', () => {
		const onItemAdd = jest.fn();
		const customEvent1 = { preventDefault: () => {}, charCode: 10, target: { value: '1' } };
		const customEvent2 = { preventDefault: () => {}, charCode: 20, target: { value: '2' } };
		const commaEvent = { preventDefault: () => {}, charCode: 30, target: { value: '3' } };
		const props = createTestProps({ onItemAdded: onItemAdd, charCodes: [10, 20] });
		const input = shallow(<MultipleValueTextInput {...props} />);
		input.find('input').simulate('keyPress', customEvent1);
		input.find('input').simulate('keyPress', customEvent2);
		input.find('input').simulate('keyPress', commaEvent);
		expect(onItemAdd).toHaveBeenCalledTimes(2);
	});
	it('should handle blur to add values when applicable', () => {
		const onItemAdd = jest.fn();
		const customEvent1 = { preventDefault: () => {}, target: { value: 'test' } };
		const props = createTestProps({ onItemAdded: onItemAdd, shouldAddOnBlur: true });
		const input = shallow(<MultipleValueTextInput {...props} />);
		input.find('input').simulate('blur', customEvent1);
		expect(onItemAdd).toHaveBeenCalledTimes(1);
		expect(onItemAdd).toHaveBeenLastCalledWith('test', ['test']);
	});
	it('should ignore blur event when not applicable', () => {
		const onItemAdd = jest.fn();
		const customEvent1 = { preventDefault: () => {}, target: { value: 'test' } };
		const props = createTestProps({ onItemAdded: onItemAdd, shouldAddOnBlur: false });
		const input = shallow(<MultipleValueTextInput {...props} />);
		input.find('input').simulate('blur', customEvent1);
		expect(onItemAdd).toHaveBeenCalledTimes(0);
	});
	it('should update state on add', () => {
		const enterEvent = { preventDefault: () => {}, charCode: 13, target: { value: 'abc' } };
		const props = createTestProps({ values: ['1', '2'] });
		const input = shallow(<MultipleValueTextInput {...props} />);
		input.find('input').simulate('change', { target: { value: 'abc' } });
		input.find('input').simulate('keyPress', enterEvent);
		expect(input.find('input')).toHaveProp('value', '');
		expect(input.find(MultipleValueTextInputItem)).toHaveLength(3);
	});
	it('should ignore duplicates', () => {
		const enterEvent = { preventDefault: () => {}, charCode: 13, target: { value: '1' } };
		const props = createTestProps({ values: ['1', '2'] });
		const input = shallow(<MultipleValueTextInput {...props} />);
		input.find('input').simulate('change', { target: { value: '1' } });
		input.find('input').simulate('keyPress', enterEvent);
		expect(input.find('input')).toHaveProp('value', '');
		expect(input.find(MultipleValueTextInputItem)).toHaveLength(2);
	});
	it('should call external remove method', () => {
		const onItemDelete = jest.fn();
		const props = createTestProps({ onItemDeleted: onItemDelete, values: ['1', '2'] });
		const input = shallow(<MultipleValueTextInput {...props} />);
		input.instance().handleItemRemove('1');
		expect(onItemDelete).toHaveBeenCalledTimes(1);
	});
	it('should remove items from state', () => {
		const props = createTestProps({ values: ['1', '2'] });
		const input = shallow(<MultipleValueTextInput {...props} />);
		input.instance().handleItemRemove('1');
		input.update();
		expect(input.find(MultipleValueTextInputItem)).toHaveLength(1);
	});
});
