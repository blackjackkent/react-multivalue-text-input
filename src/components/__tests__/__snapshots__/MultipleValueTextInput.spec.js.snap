// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should render valid snapshot with default props 1`] = `
<div
  className="multiple-value-text-input"
>
  <label
    className=""
    htmlFor="test-input"
  >
    <div
      className="multiple-value-text-input-item-container"
    />
    <input
      className=""
      name="test-input"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder=""
      type="text"
      value=""
    />
  </label>
</div>
`;

exports[`rendering should render valid snapshot with passed props 1`] = `
<div
  className="multiple-value-text-input"
>
  <label
    className=""
    htmlFor="test-input"
  >
    Test Label
    <div
      className="multiple-value-text-input-item-container"
    />
    <input
      className=""
      name="test-input"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Test Placeholder"
      type="text"
      value=""
    />
  </label>
</div>
`;
